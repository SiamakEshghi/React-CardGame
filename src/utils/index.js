import clubIcon from 'img/clubIcon.png';
import diamondIcon from 'img/diamondIcon.png';
import heartIcon from 'img/heartIcon.png';
import spadeIcon from 'img/spadeIcon.png';

const SUITES = ['heart', 'club', 'diamond', 'spade'];
const VALUES = ['ACE', '2', '3', '4', '5', '6', '7', '8', '9', '10', "J", "Q", "K"];

export const getIcon = (suitName) => {
    switch (suitName) {
        case 'club' : return clubIcon;
        case 'diamond' : return diamondIcon;
        case 'heart' : return heartIcon;
        case 'spade' : return spadeIcon;
        default: return
    }
}

export const makeDeck = () => {
    let deck = [];
    for (let i in SUITES) {
        for(let j in VALUES) {
            deck = [...deck, {suite: SUITES[i], value: VALUES[j]}]
        }
    }

    return deck;
}