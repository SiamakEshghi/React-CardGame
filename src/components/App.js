import React, { PureComponent, Fragment } from 'react';
import Header from 'components/Header';
import Card from 'components/Card';
import 'components/style.css';
import { makeDeck } from 'utils';


class App extends PureComponent {
    constructor (props) {
        super(props);
        this.state = {
            deck: [],
            dealedDeck: [],
            counter: 0,
            loading: false
        }
    }
    componentDidMount (){
        this.setState({ deck: makeDeck() })
    }
    dealCard = () => {
        const { loading, counter, deck, dealedDeck } = this.state;

        if(!loading && (counter < deck.length)) {
            this.setState({ dealedDeck: [...dealedDeck, deck[counter]]},
                () => this.setState({ counter: counter + 1 })
            );
        } 
    }
    handleShuffle = () => {
        this.setState({ 
            loading: true,
            dealedDeck: [],
            counter: 0
            }, this.shuflleCard)
    }
    shuflleCard = () => {
        const { deck } = this.state;
        const shuffleDeck = deck;
        for(let i in shuffleDeck) {
            const randIndex = Math.floor(Math.random() * (51));
            const randCard = shuffleDeck[randIndex];
            shuffleDeck[randIndex] = shuffleDeck[i];
            shuffleDeck[i] = randCard ;
        }
        this.setState({ deck: shuffleDeck}, () => {
            //Do nothing just Get time  for showing loading
            setTimeout(() => this.setState({ loading: false }), 1000)
        })
    }
    renderDeckZone = () => {
        const { deck, counter, loading } = this.state;
        return (
            <div className="deckZone" >
                <h4>Click On Deck For Deal</h4>
                <div className="dealBtn" onClick={this.dealCard}>
                    {
                    loading ? <Card loading />
                    : counter < deck.length 
                    ? <Card back={true} />
                    : <Card />
                    }
                </div>
                <button onClick={this.handleShuffle}>Shuffle</button>
            </div>
        );
    }
    renderDealedcard = () => {
        const { dealedDeck } = this.state;
        return dealedDeck.map((deck, index) => 
            <div className="dealedDeck" key={index}>
                <Card value={deck.value} suite={deck.suite} />
            </div>
        );
    }
    render() {
        return(
            <div>
                <Header />
                <div className="container">
                        {this.renderDeckZone()}
                        <div className="dealedList">
                            {this.renderDealedcard()}
                        </div>
                </div>
            </div>
        );
    }
}
export default App;