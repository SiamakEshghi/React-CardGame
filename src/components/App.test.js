import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from 'components/App';
import { makeDeck } from 'utils';

let wrapper;

beforeEach(() => {
    wrapper = mount(<App />);
    // console.log(wrapper.debug());
});

afterEach(() => {
    wrapper.unmount();
})
describe('App UI', () => {
    it('Should render correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    })
    it('When we run App it should show only one card', () => {
        expect(wrapper.find('.back').length).toEqual(1);
    })
    it('Should show a Cylon loading if state.loading is true', () => {
        wrapper.setState({ loading: true });
        wrapper.update();
        expect(wrapper.find('Loading').length).toEqual(1);
    });
    it('Should show no more card when counter is equal to 52', () => {
        wrapper.setState({ counter: 52 });
        wrapper.update();
        expect(wrapper.render().text()).toContain('No More Card!');
    });
});

describe('Deal Action', () => {
    it('Should show a new dealed card after each click', () => {
        wrapper.find('.dealBtn').simulate('click');
        wrapper.find('.dealBtn').simulate('click');
        wrapper.update();
        expect(wrapper.find('.dealedDeck').length).toEqual(2);
    });
    it('Should show add a new card to dealedDeck after each click', () => {
        wrapper.find('.dealBtn').simulate('click');
        wrapper.find('.dealBtn').simulate('click');
        wrapper.update();
        expect(wrapper.state('dealedDeck').length).toEqual(2);
    })
});

describe('Shuffle Action', () => {
    beforeEach(() => {
        wrapper.find('.dealBtn').simulate('click');
        wrapper.find('.dealBtn').simulate('click');
        wrapper.find('button').simulate('click');
        wrapper.update();
    })
    it('Deck order should be different after shuffle', () =>{
        const firstdeck = makeDeck();
        const secondDeck = wrapper.state('deck');
        expect(firstdeck[0]).not.toEqual(secondDeck[0]);
    });

    it('should remode all dealed card after shuffle', () => {
        expect(wrapper.state('dealedDeck').length).toEqual(0);
    })
});