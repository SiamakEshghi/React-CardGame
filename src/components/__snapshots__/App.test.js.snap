// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App UI Should render correctly 1`] = `
<App>
  <div>
    <Component>
      <div
        className="header"
      >
        <img
          className="logo"
          src="logo.png"
        />
        <h2>
          CARD GAME
        </h2>
        <h5>
          React App
        </h5>
      </div>
    </Component>
    <div
      className="container"
    >
      <div
        className="deckZone"
      >
        <h4>
          Click On Deck For Deal
        </h4>
        <div
          className="dealBtn"
          onClick={[Function]}
        >
          <Component
            back={true}
          >
            <div
              className="card"
            >
              <img
                className="back"
                src="cardBack.png"
              />
            </div>
          </Component>
        </div>
        <button
          onClick={[Function]}
        >
          Shuffle
        </button>
      </div>
      <div
        className="dealedList"
      />
    </div>
  </div>
</App>
`;
